sound/soc/codecs/rt1318-sdw.c

```diff
@@ -0,0 +1,955 @@
+	case 0xea01:
+	case 0xebc5:
+	case 0xebc8:
+	case 0xebcb ... 0xebcc:
+	case 0xed03 ... 0xed06:
+	case 0xf010 ... 0xf014:
+	case SDW_SDCA_CTL(FUNC_NUM_SMART_AMP, RT1318_SDCA_ENT_SAPU, RT1318_SDCA_CTL_SAPU_PROTECTION_MODE, 0):
+	case SDW_SDCA_CTL(FUNC_NUM_SMART_AMP, RT1318_SDCA_ENT_SAPU, RT1318_SDCA_CTL_SAPU_PROTECTION_STATUS, 0):
+		return true;
+	default:
+		return false;
+	}
+}
+
+static const struct regmap_config rt1318_sdw_regmap = {
+	.reg_bits = 32,
+	.val_bits = 8,
+	.readable_reg = rt1318_readable_register,
+	.volatile_reg = rt1318_volatile_register,
+	.max_register = 0x41081488,
+	.reg_defaults = rt1318_reg_defaults,
+	.num_reg_defaults = ARRAY_SIZE(rt1318_reg_defaults),
+	.cache_type = REGCACHE_RBTREE,
+	.use_single_read = true,
+	.use_single_write = true,
+};
+
+static int rt1318_read_prop(struct sdw_slave *slave)
+{
+	struct sdw_slave_prop *prop = &slave->prop;
+	int nval;
+	int i, j;
+	u32 bit;
+	unsigned long addr;
+	struct sdw_dpn_prop *dpn;
+
+	prop->scp_int1_mask = SDW_SCP_INT1_BUS_CLASH | SDW_SCP_INT1_PARITY;
+	prop->quirks = SDW_SLAVE_QUIRKS_INVALID_INITIAL_PARITY;
+	prop->is_sdca = true;
+
+	prop->paging_support = true;
+
+	/* first we need to allocate memory for set bits in port lists */
+	prop->source_ports = BIT(2);
+	prop->sink_ports = BIT(1);
+
+	nval = hweight32(prop->source_ports);
+	prop->src_dpn_prop = devm_kcalloc(&slave->dev, nval,
+		sizeof(*prop->src_dpn_prop), GFP_KERNEL);
+	if (!prop->src_dpn_prop)
+		return -ENOMEM;
+
+	i = 0;
+	dpn = prop->src_dpn_prop;
+	addr = prop->source_ports;
+	for_each_set_bit(bit, &addr, 32) {
+		dpn[i].num = bit;
+		dpn[i].type = SDW_DPN_FULL;
+		dpn[i].simple_ch_prep_sm = true;
+		dpn[i].ch_prep_timeout = 10;
+		i++;
+	}
+
+	/* do this again for sink now */
+	nval = hweight32(prop->sink_ports);
+	prop->sink_dpn_prop = devm_kcalloc(&slave->dev, nval,
+		sizeof(*prop->sink_dpn_prop), GFP_KERNEL);
+	if (!prop->sink_dpn_prop)
+		return -ENOMEM;
+
+	j = 0;
+	dpn = prop->sink_dpn_prop;
+	addr = prop->sink_ports;
+	for_each_set_bit(bit, &addr, 32) {
+		dpn[j].num = bit;
+		dpn[j].type = SDW_DPN_FULL;
+		dpn[j].simple_ch_prep_sm = true;
+		dpn[j].ch_prep_timeout = 10;
+		j++;
+	}
+
+	/* set the timeout values */
+	prop->clk_stop_timeout = 20;
+
+	return 0;
+}
+
+static int rt1318_spk_coeff_init(struct rt1318_sdw_priv *rt1318)
+{
+	int ret = 0, set_flag;
+	char filename[128];
+	const char *vendor, *product;
+	struct device *dev = &rt1318->sdw_slave->dev;
+	const struct firmware *fw = NULL;
+	char *param_ptr;
+	unsigned int i, reg, data;
+
+	ret = regmap_read(rt1318->regmap, 0x3000, &set_flag);
+	if (!set_flag) {
+
+		if (rt1318->param_name == NULL) {
+			vendor = dmi_get_system_info(DMI_SYS_VENDOR);
+			product = dmi_get_system_info(DMI_PRODUCT_SKU);
+			dev_dbg(dev, "vendor:%s, product:%s\n", vendor, product);
+
+			if (vendor && product)
+				snprintf(filename, sizeof(filename),
+					"rt1318_param_%s_%s.bin", vendor, product);
+			else
+				sprintf(filename, "rt1318_param.bin");
```

Comment:
@plbossart Because RT1318 has the hardware speaker protection function, that has a lot of setting to enable it.
Therefore, I want to use the binary file to store the settings for the different models.
I could remove this function. Dell doesn't request to enable it in Linux for now.

Reply:
I think it's fine to use the firmware, I am just worried about using default settings that may actually do more harm than good.

Could we enable this only for specific/known SKUs and if the SKU is not known then the protection is disabled and we choose more conservative settings.

Reply:
@plbossart Ok, got it. If the SKU wants to enable the protection function, I will add this function back and check the specific model to call this function.